versions/3.0.4.md

```diff
@@ -2302,7 +2301,7 @@ For computing links and providing instructions to execute them, a [runtime expre
 
 | Field Name | Type | Description |
 | ---- | :----: | ---- |
-| <a name="link-operation-ref"></a>operationRef | `string` | A relative or absolute URI reference to an OAS operation. This field is mutually exclusive of the `operationId` field, and MUST point to an [Operation Object](#operation-object). Relative `operationRef` values MAY be used to locate an existing [Operation Object](#operation-object) in the OpenAPI description. |
+| <a name="link-operation-ref"></a>operationRef | `string` | A URI reference to an OAS operation. This field is mutually exclusive of the `operationId` field, and MUST point to an [Operation Object](#operation-object). Relative `operationRef` values MAY be used to locate an existing [Operation Object](#operation-object) in the OpenAPI Description. |
```

Comment:
```suggestion
| <a name="link-operation-ref"></a>operationRef | `string` | A URI reference to an OAS operation. This field is mutually exclusive of the `operationId` field, and MUST point to an [Operation Object](#operation-object). |
```
Can we remove the confusing last sentence? The URI reference MUST point to an operation object, that's it. This implies that the referenced thingy exists, otherwise it can't be an Operation Object.

Reply:
Maybe not. In the Relative References in URLs section it says:

> Unless specified otherwise, all fields that are URLs MAY be relative references as defined by [RFC3986](https://tools.ietf.org/html/rfc3986#section-4.2).
> Relative references are resolved using the URLs defined in the [Server Object](#server-object) as a Base URI.
> 
> Relative references used in `$ref` are processed as per [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03), using the URL of the current document as the base URI. See also the [Reference Object](#reference-object).

I think this sentence is "specifying otherwise" -- that operation refs are _not_ resolved using the URLs in the Server Object as the Base URL, but instead using the "OpenAPI Description" (current document). Okay, that is confusing but I think resolving using the URLs in the Server Object is really not what we want.

Reply:
This sentence says nothing about _resolving_ the relative URI and especially says nothing about the base URI to be used.

Reply:
On closer inspection the sentence adds nothing that hasn't already been said in the two preceding sentences.

Reply:
I agree that the sentence is confusing and may not add anything in its current form. But I am suggesting that we need to replace it with a sentence that explains that URI references are resolved based on the current document.  Maybe:
> Relative references used in this field are processed as per [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03), using the URL of the current document as the base URI. See also the [Reference Object](https://github.com/OAI/OpenAPI-Specification/pull/4100#reference-object).

or something shorter, like

> Relative references in this field are processed in the same manner as in the [Reference Object](https://github.com/OAI/OpenAPI-Specification/pull/4100#reference-object).

Reply:
@mikekistler 3.0 only defines one process to determine a base URI, so that MUST apply to all URI-reference resolutions.  That process is:

1.  Nearest Server Object `url` field (under RFC3986 §5.1.1)
2. (we act like §5.1.2 doesn't exist, but it should take precedence at this point)
3. Retrieval URL of the current document (§5.1.3)
4. (we don't define an application-specific default under §5.1.4, leaving that instead to OAS implementations, which is an appropriate way to handle this)

Reply:
@handrews Perhaps I am misreading/misunderstanding this sentence in [Relative References in URLs](https://github.com/OAI/OpenAPI-Specification/blob/v3.0.4-dev/versions/3.0.4.md#relative-references-in-urls):

> Relative references used in $ref are processed as per [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03), using the URL of the current document as the base URI. See also the [Reference Object](https://github.com/OAI/OpenAPI-Specification/pull/4100#reference-object).

Doesn't this mean that $ref values use the URL of the current document as the base URI?

Reply:
Right. So isn't the same true of `operationRef` objects?

Reply:
@mikekistler I'm a little confused.  In 3.0.x, these four steps for finding the base URI:

1.  Nearest Server Object `url` field (under RFC3986 §5.1.1)
2. (we act like §5.1.2 doesn't exist, but it should take precedence at this point)
3. Retrieval URL of the current document (§5.1.3)
4. (we don't define an application-specific default under §5.1.4, leaving that instead to OAS implementations, which is an appropriate way to handle this)

apply to _all_ relative URI-references (which in 3.0 are all treated as URL-references).

The note about the JSON Reference specification does not change this for anything.

Reply:
I _think_ the difference is that you're jumping to step 3 for `operationRef`?  Or did I misread?  Which is entirely possible, I'm just popping in and out of these conversations while traveling this past week and might have missed something.

Reply:
Actually I am taking [4.6 Relative References in URLs](https://spec.openapis.org/oas/v3.0.3.html#relative-references-in-urls) and the statement

> Relative references used in $ref are processed [...] using the URL of the current document as the base URI. See also the [Reference Object](https://spec.openapis.org/oas/v3.0.3.html#reference-object).

quite literally and _never_ consider a Server Object for relative URL resolution in `$ref`, always starting with your step 

> 3. Retrieval URL of the current document (RFC3986 §5.1.3)

Which is also how the quite popular OpenAPI visualization tool that I am using works, so changing this in 3.0.4 would break some implementations.

Reply:
@ralfhandl OK so you read that section as having two completely separate requirements?  I had never read it that way, possibly because I read RFC3986 as taking precedence over wording in either JSON Reference or OAS that seems to disregard how RFC3986 is supposed to work.

_[The below section is a bit of a rant if you're curious as to how I read it, but I'm basically giving up on my reading so feel free to skip to the next comment, and don't worry if it doesn't make sense to you.]_

JSON Schema draft-04 _also_ referenced JSON Reference, and modified it by adding `id` as a base URI source per RFC3986 §5.1.1.  To me, the only sensible way to read OAS 3.0 is that it does the exact same thing as JSON Schema draft-04:  It inserts an RFC3986 §5.1.1-compliant source for the base URI from within the OAD contents.  This usage of the Server Object's `url` field _exactly_ parallels JSON Schema draft-04's usage of `id`.

But I also read both the OAS and JSON Reference as _invalidly_ trying to short-circuit RFC3986 by jumping to §5.1.3.  The OAS also reads very strangely as it seems to try to contrast the Reference Object with JSON Schema behavior even though there is no possible reading of the Reference Object that would have it behave like JSON Schema. I started trying to explain exactly why and how I see that but, really, I don't care how 3.0 resolves things at this point as long as it is clear.

Reply:
@ralfhandl @mikekistler If we accept Ralf's reading, which appears to also be Mike's reading, then we definitely have a problem because this was written without thinking through the fields involved.  Here are the places where we have relative URI-references in 3.0 ("retrieval URL" is per RFC3986 §5.1.3 and means the current document's retrieval URL — I'm leaving out any mention of §5.1.2 or §5.1.4 because the OAS and JSON Reference improperly fail to mention them):

| Object | field | base URI search per text | Notes
| ------- | ---- | --------- | ----
| Contact | `url` | root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| License | `url` | root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| Server | `url` | retrieval URL | _the Server object `url` can also be relative_ |
| Paths | `/{path}` | operation&nbsp;Server&nbsp;Object&nbsp;`url`; path&nbsp;item&nbsp;Server&nbsp;Object&nbsp;`url`; root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| Path Item | `$ref` | retrieval URL | _I assume? Because §4.6 says "`$ref`" and not Reference Object?_ |
| External Documentation | `url` | root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| Example | `externalValue` | operation&nbsp;Server&nbsp;Object&nbsp;`url`; path&nbsp;item&nbsp;Server&nbsp;Object&nbsp;`url`; root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| Link | `operationRef` | root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| Reference | `$ref` | retrieval URL |
| Discriminator | `mapping` | root&nbsp;Server&nbsp;Object&nbsp;`url`; retrieval&nbsp;URL |
| _any_ | `description` markdown | base URI of rendered documentation resource | _clarified in 3.0.4, previously ambiguous_ |

This is rather chaotically inconsistent, which is part of why I always read it differently.  The above behavior does not make sense to me, but it is the logical conclusion of your insistence that `$ref` (without qualification for _which_ `$ref`) is treated differently from everything else.  It's also frustrating because in 3.1 we defined a consistent URL vs URI split, and it does _not_ match the split that this reading causes in 3.0.

I'm not sure there's anything that can be done about this if what @ralfhandl says is indeed how it has always been interpreted.  But we should call out exactly what the split is (just as we do in 3.1, but without the help of the URI vs URL distinction) because I don't think this is obvious at all.

Reply:
@ralfhandl @mikekistler I suppose we _could_ just declare that the intent was the same sort of division later clarified as URIs vs URLs in 3.1, and _not_ literally specific to `$ref`, which would produce behavior like this:


| Object | field | base URI search order
| ------- | ---- | ---------
| Contact | `url` | retrieval URL |
| License | `url` | retrieval URL |
| Server | `url` | retrieval URL |
| Paths | `/{path}` | link&nbsp;Server&nbsp;Object&nbsp;`url`<br />operation&nbsp;Server&nbsp;Object&nbsp;`url`<br />path&nbsp;item&nbsp;Server&nbsp;Object&nbsp;`url`<br />root&nbsp;Server&nbsp;Object&nbsp;`url`<br />retrieval&nbsp;URL |
| Path Item | `$ref` | retrieval URL |
| External Documentation | `url` | retrieval URL |
| Example | `externalValue` | retrieval URL |
| Link | `operationRef` | retrieval URL |
| Reference | `$ref` | retrieval URL |
| Discriminator | `mapping` | retrieval URL |
| _any_ | `description` markdown | base URI of rendered documentation resource | _clarified in 3.0.4, previously ambiguous_ |

Also, in these tables I'm missing the URLs in the Security Scheme and OAuth Flow Objects, which in this table would still check the root Server Obect prior to falling back to the retrieval URL.

Reply:
That nicely sums up my understanding of how link resolution in 2.0 and 3.0.x is supposed to work!

Reply:
We agreed to make one more minor change here, this won't cause a problem in 3.1 so we need to clarify in the spec to make the second table's meaning clearly what is intended.

Reply:
After a conversation with Henry, I have come to agree with the latest table as our recommended behaviour.  We will need to add some text in the relative references section to clarify that the current wording does not match with this behaviour and tooling may or may not follow the recommendations.